### PR TITLE
enhancement(graphql api): Refactor `outputEvents` subscription -> `outputEventsByComponentIdPatterns`

### LINKED ISSUE #9317: Investigate tap API notification change for (un)matched components
The tap API notifies when provided component filter patterns match (or don't match) underlying topology.

Historically, the alerts echoed the glob pattern used to filter:

https://github.com/vectordotdev/vector/blob/e06248659c5fa55e8829562295c4522cc058692d/src/api/tap.rs#L51-L59

This has been updated to use `ComponentKey::from()`, which feels like an error, since deriving a definitive component from a wildcard isn't possible:

https://github.com/vectordotdev/vector/blob/593cfe5d35cd0215ca0e35501288a32708eeedda/src/api/tap.rs#L53-L59

This could do with auditing to ensure the correct notifications are being sent back, and we're not attempting to unmarshal a glob into something definitive.

### PR DESCRIPTION
**Pull request by leebenson**
This PR introduces a breaking change in the tap API that aims to fix the erroneous fields covered in #9317.

Closes #9317.

## Previous issue

The signature of the tap subscription prior to this change was:

```gql
subscription {
  outputEvents(componentIds:["bla*"]) {
    ... on EventNotification {
      componentId
      pipelineId
      notification
    }
  }
}
```

Returning (for example):

```json{
  "data": {
    "outputEvents": [
      {
        "componentId": "bla*",
        "pipelineId": null,
        "notification": "NOT_MATCHED"
      }
    ]
  }
}
```

There are several issues with this signature:

1. The `componentIds` param in the subscription isn't actually a list of component IDs, but rather glob patterns that can expand internally to IDs
2. The purpose of the matched/not matched notifications is to echo the _patterns_ that matched; not a definitive componentId and pipelineId as the type would suggest.
3. The change was introduced alongside the pipelines feature, which itself could represent a breaking change to the `EventNotification` type. Therefore, I deem _another_ breaking change that attempts to clarify the API signature to be relatively low-risk.
4. Internally, `ComponentKey::from` is being used to attempt to unmarshal a pattern back to a definitive component. This generally won't succeed (at least in the way intended) as the returned _componentId_ is likely to be `*` for most uses cases in tap, and the _pipelineId_ will be null -- even if the component itself belongs to a pipeline.

## Suggested change

This PR removes the previous signature in favour of:

```gql
subscription {
  outputEventsByComponentIdPatterns(patterns:["bla*"]) {
    ... on EventNotification {
      pattern
      notification
    }
  }
}
```

Returning (for example):

```json
{
  "data": {
    "outputEventsByComponentIdPatterns": [
      {
        "notification": "NOT_MATCHED",
        "pattern": "bla*"
      }
    ]
  }
}
```

Changes:

1. The field name has been renamed from `outputEvents` -> `outputEventsByComponentIdPatterns`, to reflect that the parameter it takes is explicitly scoped to component ID patterns. With the introduction of pipeline IDs, there was ambiguity introduced in the API that made it how components would be resolved if, for example, a separate `pipelineIds` param was introduced for filtering.
2. The param for patterns has been renamed from `componentIds` to `patterns`, to reflect the field context.
3. `EventNotification` returns a `pattern` field instead of separate _componentId_ and _pipelineId_ fields.

## Breaking change

The rationale for introducing a breaking change is as follows:

1. The existing types were erroneous and suggested a use-case contrary to what the `EventNotification` type was intended to communicate. If left, it could cause false alerts/confusion for users.
2. The API isn't widely publicised and for all practical purposes is an internal implementation detail that isn't likely to be relied upon outside of `vector tap` currently.
3. We lack an official deprecation/support policy which may provide some padding against internal API changes until we guarantee versioning.
4. This changes both the server and `vector tap` client, ensuring a released version has compatibility between the client and the server it observes.

With that said, this _will_ break compatibility between newer and older clients. Note that this may have been broken anyway due to recent pipeline changes in the API. I think this is a good opportunity to clarify our client/server compatibility policy between versions in any case.

### CODE REVIEW

File: lib/vector-api-client/src/gql/tap.rs [hunk 1]
```diff
@@ -47,45 +49,46 @@ impl From<TapEncodingFormat> for output_events_subscription::EventEncodingType {
     }
 }
 
-impl output_events_subscription::OutputEventsSubscriptionOutputEvents {
+impl output_events_by_component_id_patterns_subscription::OutputEventsByComponentIdPatternsSubscriptionOutputEventsByComponentIdPatterns {
     pub fn as_log(
         &self,
-    ) -> Option<&output_events_subscription::OutputEventsSubscriptionOutputEventsOnLog> {
+    ) -> Option<&output_events_by_component_id_patterns_subscription::OutputEventsByComponentIdPatternsSubscriptionOutputEventsByComponentIdPatternsOnLog>{
```
Comment (jszwedko) [inline lib/vector-api-client/src/gql/tap.rs:55]: This name 😂

File: lib/vector-api-client/src/gql/tap.rs [hunk 2]
```diff
@@ -47,45 +49,46 @@ impl From<TapEncodingFormat> for output_events_subscription::EventEncodingType {
...
     }
 }
 
 pub trait TapSubscriptionExt {
     /// Executes an output events subscription.
-    fn output_events_subscription(
+    fn output_events_by_component_id_patterns_subscription(
         &self,
-        component_ids: Vec<String>,
+        component_patterns: Vec<String>,
```
Comment (jszwedko) [inline lib/vector-api-client/src/gql/tap.rs:67]: 👍 this naming should make it clearer.

**Review by jszwedko — APPROVED** [review]
Thanks for fixing this @leebenson . The modeling of patterns seems clearer now and should help avoid this sort of misunderstanding in the future.

**Review by lukesteensen — COMMENTED** [review]
Looks good! Thanks for handling this @leebenson

**Pull request merged**